src/cuda-kernels/allocators.cu

```diff
@@ -0,0 +1,326 @@
+                                          sizeof(T) * length);
+  if(malloc_stat != cudaError_t::cudaSuccess) {
+      return failure(cudaGetErrorString(malloc_stat), 0, kSliceNone, true);
+  }
+  cudaError_t memcpy_stat = cudaMemcpy(*to_ptr,
+                                       from_ptr,
+                                       sizeof(T) * length,
+                                       cudaMemcpyHostToDevice);
+
+  if(memcpy_stat != cudaError_t::cudaSuccess) {
+      return failure(cudaGetErrorString(memcpy_stat), 0, kSliceNone, true);
+  }
+  return success();
+}
+ERROR awkward_cuda_H2Dbool(
+  bool **to_ptr,
+  bool *from_ptr,
+  int64_t length) {
+  return awkward_cuda_H2D<bool>(
+    to_ptr,
+    from_ptr,
+    length);
+}
+ERROR awkward_cuda_H2D8(
+  int8_t **to_ptr,
+  int8_t *from_ptr,
+  int64_t length) {
+  return awkward_cuda_H2D<int8_t>(
+    to_ptr,
+    from_ptr,
+    length);
+}
+ERROR awkward_cuda_H2DU8(
+  uint8_t **to_ptr,
+  uint8_t *from_ptr,
+  int64_t length) {
+  return awkward_cuda_H2D<uint8_t>(
+    to_ptr,
+    from_ptr,
+    length);
+}
+ERROR awkward_cuda_H2D16(
+  int16_t **to_ptr,
+  int16_t *from_ptr,
+  int64_t length) {
+  return awkward_cuda_H2D<int16_t>(
+    to_ptr,
+    from_ptr,
+    length);
+}
+ERROR awkward_cuda_H2DU16(
+  uint16_t **to_ptr,
+  uint16_t *from_ptr,
+  int64_t length) {
+  return awkward_cuda_H2D<uint16_t>(
+    to_ptr,
+    from_ptr,
+    length);
+}
+ERROR awkward_cuda_H2D32(
+  int32_t **to_ptr,
+  int32_t *from_ptr,
+  int64_t length) {
+  return awkward_cuda_H2D<int32_t>(
+    to_ptr,
+    from_ptr,
+    length);
+}
+ERROR awkward_cuda_H2DU32(
+  uint32_t **to_ptr,
+  uint32_t *from_ptr,
+  int64_t length) {
+  return awkward_cuda_H2D<uint32_t>(
+    to_ptr,
+    from_ptr,
+    length);
+}
+Error awkward_cuda_H2D64(
+  int64_t **to_ptr,
+  int64_t *from_ptr,
+  int64_t length) {
+  return awkward_cuda_H2D<int64_t>(
+    to_ptr,
+    from_ptr,
+    length);
+}
+ERROR awkward_cuda_H2DU64(
+  uint64_t **to_ptr,
+  uint64_t *from_ptr,
+  int64_t length) {
+  return awkward_cuda_H2D<uint64_t>(
+    to_ptr,
+    from_ptr,
+    length);
+}
+ERROR awkward_cuda_H2Dfloat32(
+  float **to_ptr,
+  float *from_ptr,
+  int64_t length) {
+  return awkward_cuda_H2D<float>(
+    to_ptr,
+    from_ptr,
+    length);
+}
+ERROR awkward_cuda_H2Dfloat64(
+  double **to_ptr,
+  double *from_ptr,
+  int64_t length) {
+  return awkward_cuda_H2D<double>(
+    to_ptr,
+    from_ptr,
+    length);
+}
+
+template <typename T>
+ERROR awkward_cuda_D2H(
+        T** to_ptr,
+        T* from_ptr,
+        int64_t length) {
+    cudaError_t memcpy_stat = cudaMemcpy(*to_ptr,
+                                         from_ptr,
+                                         sizeof(T) * length,
+                                         cudaMemcpyDeviceToHost);
+
+    if(memcpy_stat != cudaError_t::cudaSuccess) {
+        return failure(cudaGetErrorString(memcpy_stat), 0, kSliceNone, true);
```

Comment:
So the `extra` passes a message through without modification if `true`. Maybe a good name would be `pass_through`.